framework/curl_client.py

```diff
@@ -0,0 +1,310 @@
+            match = re.match(r"HTTP/([.12]+) (\d+)", response_line)
+            self.proto = match.group(1)
+            self.status = int(match.group(2))
+
+            self._multi_headers = defaultdict(list)
+            for k, v in ((k.lower(), v) for k, v in message.items()):
+                self._multi_headers[k].append(v)
+            self.headers = {k: v[-1] for k, v in self.multi_headers.items()}
+
+
+@dataclass
+class CurlArguments:
+    """Interface class for cURL client.
+    Contains all accepted arguments (fields) supported by `CurlClient`.
+    """
+
+    addr: str
+    uri: str = "/"
+    cmd_args: str = ""
+    data: str = ""
+    headers: dict = None
+    dump_headers: int = True
+    disable_output: bool = False
+    save_cookies: bool = False
+    load_cookies: bool = False
+    ssl: bool = False
+    http2: bool = False
+    insecure: bool = True
+    parallel: int = None
+
+    @classmethod
+    def get_kwargs(cls) -> List[str]:
+        """Returns list of `CurlClient` supported argument names."""
+        return list(cls.__dataclass_fields__.keys())
+
+
+class CurlClient(CurlArguments, client.Client):
+    """
+    Wrapper to manage cURL.
+    See `selftests/test_curl_client.py` and #332 PR for usage examples.
+
+    Args:
+      addr (str): Host (IP address) to connect to
+      uri (str): URI to access
+      cmd_args (str): additional curl options
+      data (str): data to POST
+      headers (dict[str, str]): headers to include in the request
+      dump_headers (bool): dump headers to the workdir, and enable response parsing.
+                           Enabled by default.
+      disable_output (bool): do not output results but only errors
+      save_cookies (bool): save cookies to the workdir
+      load_cookies (bool): load and send cookies from the workdir
+      ssl (bool): use SSL/TLS for the connection
+      http2 (bool): use HTTP version 2
+      insecure (bool): Ignore SSL certificate errors. Enabled by default.
+      parallel (int): Enable parallel mode, with maximum <int> simultaneous transfers
+    """
+
+    def __init__(self, **kwargs):
+        # Initialize the `CurlArguments` interface first
+        super().__init__(**kwargs)
+        # Initialize the base `Client`
+        client.Client.__init__(
+            self,
+            binary="curl",
+            server_addr=self.addr,
+            uri=self.uri,
+            ssl=self.ssl or self.http2,
+        )
+        self.options = [self.cmd_args] if self.cmd_args else []
+        self._responses = []
+        self._stats = []
+        self._statuses = defaultdict(lambda: 0)
+        self._output_delimeter = "-===curl-transfer===-"
+
+    @property
+    def responses(self) -> List[CurlResponse]:
+        """List of all received responses."""
+        return list(self._responses)
+
+    @property
+    def last_response(self) -> Optional[CurlResponse]:
+        """Last parsed response if any."""
+        return (self.responses or [None])[-1]
```

Comment:
I think this method should be changed. For test: 
```
    def test_error_on_wrong_port(self):
        client: CurlClient = self.get_client("wrong_port")
        client.uri = '127.0.0.1:80'
        self.get_response(client)
        client.uri = '127.0.0.1:443'
        response = self.get_response(client)
        self.assertFalse(response)
```
I receive last response to the past request:
```
Traceback (most recent call last):
  File "/home/cookies/tempesta-test/selftests/test_curl_client.py", line 205, in test_error_on_wrong_port
    self.assertFalse(response)
AssertionError: CurlResponse(status=200, proto='1.1', headers={'date': 'test', 'x-header': 'Test-Value-2', 'set-cookie': 'curl=test', 'content-length': '4', 'via': '1.1 tempesta_fw (Tempesta FW pre-0.7.0)', 'server': 'Tempesta FW/pre-0.7.0'}) is not false
```
There is no need to link last_response to list of responses.

Reply:
Test should now work as expected, the `responses` list is now cleared between restarts.